wallet: Fix unique_ptr usage in boost::signals2

promag opened the pull request:
This PR includes 2 fixes:
 - prevent GUI LoadWallet handlers from crashing on startup when multiple handlers are attached, because the first handler takes ownership of the wallet unique pointer. Now every handler will receive its own unique pointer;

 - prevent showing a wallet twice in the GUI on startup due to a race with `loadwallet`.

Fixes #16937

promag commented:
@ryanofsky maybe you have a better approach?

ryanofsky commented:
Good catch! But why exactly is the fix in #16955 incorrect? It seems like a reasonable workaround to me as long as the wallet still shows up in the GUI (which I would expect if `m_node.getWallets()` is called later). But if the wallet doesn't show up in the GUI then the bigger fix is needed.

This clone fix should work now, and is more elegant than the nullptr workaround, but wouldn't work great with #10102 unless I changed it to use shared_ptrs internally or complicate lifetime management some other way to support cloning.

More ideally, I would delete `uiInterface.LoadWallet` and `Chain::loadWallet` entirely and just not use boost signals for this callback. Instead I'd define a load wallet callback list (`std::list<LoadWalletFn> load_wallet_fns`) alongside the existing list of wallets (`std::vector<std::shared_ptr<CWallet>> vpwallets`), and push onto the list when `handleLoadWallet` is called, erase from the list when `Handler::disconnect` is called, and the replace existing line:

```c++
chain.loadWallet(interfaces::MakeWallet(walletInstance));
```

with

```c++
for (auto& load_wallet : load_wallet_fns) {
    load_wallet_fn(MakeWallet(walletInstance));
}
```

But I'm ok with the nullptr fix or the clone fix or however you want to proceed here.

promag commented:
IMO it's incorrect because it doesn't remove the invalid `std::move`.

promag commented:
> which I would expect if m_node.getWallets() is called later

There's nothing guaranteeing it.

promag commented:
For reference, the bug was introduced in 1106a6fde4bfde31a16de45e4cc84ed5da05c5a4 #15288 - backport not required.

promag commented:
Related discussion https://github.com/bitcoin/bitcoin/pull/15288#discussion_r261062566.

fanquake commented:
I started playing around with this change (7162be2eb2698eeea1eaaca2159d830acca63a1e). Using the same approach as in #16937 I was able to (only once so far) end up with the same wallet loaded twice:

<img width="328" alt="duplicate_big" src="https://user-images.githubusercontent.com/863730/65687090-e8648d00-e09a-11e9-8583-a694eae9626d.png">

Multiple times I've seen `bitcoin-qt` crash during load with:
```bash
(lldb) run
Process 20318 launched: '/Users/michael/github/bitcoin/src/qt/bitcoin-qt' (x86_64)
2019-09-26 20:14:33.701294+0800 bitcoin-qt[20318:1442147] MessageTracer: Falling back to default whitelist
Process 20318 stopped
* thread #1, name = 'bitcoin-main', queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x10)
    frame #0: 0x000000010009182e bitcoin-qt`GUIUtil::TableViewLastColumnResizingFixer::disconnectViewHeadersSignals(this=0x0000000000000000) at guiutil.cpp:445:16
   442 	// We need to disconnect these while handling the resize events, otherwise we can enter infinite loops.
   443 	void TableViewLastColumnResizingFixer::disconnectViewHeadersSignals()
   444 	{
-> 445 	    disconnect(tableView->horizontalHeader(), &QHeaderView::sectionResized, this, &TableViewLastColumnResizingFixer::on_sectionResized);
   446 	    disconnect(tableView->horizontalHeader(), &QHeaderView::geometriesChanged, this, &TableViewLastColumnResizingFixer::on_geometriesChanged);
   447 	}
   448 	
Target 0: (bitcoin-qt) stopped.
```

Ran out of time tonight, so will keep looking tomorrow.

promag commented:
Thanks @fanquake!

> I was able to (only once so far) end up with the same wallet loaded twice

Very strange, I wonder if the "big" wallet has 2 `WalletModel` or not. I would suspect it is a bug on the `WalletView` creation, not on `WalletController::getOrCreateWallet` since it's guarded by a mutex.

promag committed:
gui: Fix duplicate wallet showing up

The slot BitcoinGUI::addWallet can be invoked twice for the same
WalletModel due to a concurrent wallet being loaded after the first `connect()`:

```cpp
 connect(wallet_controller, &WalletController::walletAdded, this, &BitcoinGUI::addWallet);
 connect(wallet_controller, &WalletController::walletRemoved, this, &BitcoinGUI::removeWallet);

 for (WalletModel* wallet_model : m_wallet_controller->getOpenWallets()) {
     addWallet(wallet_model);
```

promag commented:
> Multiple times I've seen bitcoin-qt crash during load with:

@fanquake that happens due to `loadwallet`? Or have you added other calls to the test script?

fanquake commented:
> that happens due to loadwallet? Or have you added other calls to the test script?

That was happening with just the `loadwallet` calls. However I have not seen either issue while testing 40660302d6a38ff0c26ecd80777990332a574c78 so far.

promag commented:
Going to `rebase --autosquash`.

ryanofsky committed:
Drop signal CClientUIInterface::LoadWallet

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/16963#issuecomment-534978002

> More ideally, I would delete `uiInterface.LoadWallet`

I haven't seen any response to this suggestion, so I pushed a quick untested implementation to https://github.com/bitcoin/bitcoin/compare/master...ryanofsky:pr/loadwallets

I think it would be a better solution than trying to clone interfaces because cloning interfaces can't work when an implementation owns noncopyable resources. I also don't agree a bug was introduced in the commit noted https://github.com/bitcoin/bitcoin/pull/16963#issuecomment-535179262 or see how #16955 is buggy or would be a bad workaround if documented appropriately.

Anyway, these are just opinions I'm giving since I was tagged earlier https://github.com/bitcoin/bitcoin/pull/16963#issuecomment-534955357. I don't feel strongly, and trust @promag's judgement on what solution to use since he's taking the initiative here.

laanwj commented:
I think calling a refactor a fix is strange: what is this, a refactor or a bug fix?

promag commented:
@ryanofsky sorry for not replying, I'll take a look.

@laanwj yeah in that case a fix.

promag commented:
@ryanofsky 

> or see how #16955 is buggy or would be a bad workaround if documented appropriately.

It's a workaround due to bad usage of std::unique_ptr and boost::signals2. I think a fix is preferable.

> I also don't agree a bug was introduced in the commit noted

What you mean? can you reproduce #16937 before 1106a6f?

> I haven't seen any response to this suggestion, so I pushed a quick untested implementation

Sweet! 5d7a233193ad2749d326607eb0e995696c8965c8 LGTM, picked and reworded.

promag commented:
Fixed disable wallet builds and squashed.

ryanofsky reviewed (approved):
utACK fce3b88ffae76d43cd5268ca364f593a5802525c, but I wrote some code in this, so discount my review a bit.

ryanofsky commented:
Should update PR description. Also it'd be nice for the PR description to summarize the fix, instead of just linking off into other issues. Maybe: "Prevent GUI LoadWallet handlers from crashing on startup when multiple handlers are attached, because the first handler takes ownership of the wallet unique pointer. Now every handler will receive its own unique pointer."

promag commented:
@ryanofsky thank you, done.

ryanofsky commented:
It looks like a rebase here got messed up, and the latest code I acked from fce3b88ffae76d43cd5268ca364f593a5802525c is gone and replaced by older clone code. I wouldn't ack the current e956882a4fc43b250e0d675dc2e0043bc94792ca without a fix or explanation.

promag commented:
Yeah forget to update local branch before rebase. Will fix, sorry.

promag commented:
@ryanofsky fixed, thanks!

ryanofsky reviewed (approved):
Code review ACK 6d6a7a8403ae923f189812edebdd95761de0e7f2. No changes since last ACK other than rebase due to #17070

ryanofsky commented:
Second commit 81ea66c30e2953dee24d5b127c28daa0d9452a28 might need more code comments and a better commit description to be more easily reviewable.

http://www.erisian.com.au/bitcoin-core-dev/log-2019-11-02.html#l-128

> \[11:33:40] <promag> wumpus: maybe see #16963?
> \[11:36:06] <wumpus> promag: I honestly don't see myself qualified to review that
> \[11:36:35] <wumpus> the inter-thread and notification logic around loading wallets is somewhat of a mystery to me

The main thing this change is doing is deleting the `boost::signals2::signal<CClientUIInterface::LoadWalletSig> LoadWallet` boost signal and replacing it with a `std::list<LoadWalletFn>` list of callback functions. In general a boost signal can be seen as equivalent to a list of callbacks. Connecting a signal is the same as adding a new callback to the list, disconnecting the signal is the same as dropping a callback from the list, and invoking the signal is the same as iterating over the list and invoking each callback.

The reason for making this change is that the load wallet callback function takes a `unique_ptr<interfaces::Wallet>` argument that is awkward to share. Problems can happen if the first callback uses `std::move` on the `unique_ptr` and later callbacks just see a null pointer. Using `std::list` instead of `boost::signal` works better because wallet load code can just pass a different `unique_ptr` to each callback, and not have to worry about moving or sharing a single value.

Another thing that might look complicated here and could be better documented is the new `MakeHandler` function:

```c++
//! Return handler wrapping a cleanup function.
std::unique_ptr<Handler> MakeHandler(std::function<void()> cleanup);
```

This just takes a `std::function<void()>` cleanup function, and returns an object which invokes the cleanup function when the object is destroyed. It's used to erase entries from the `std::list<LoadWalletFn>` list when the GUI no longer wants to receive callbacks.

promag commented:
> This just takes a `std::function<void()>` cleanup function, and returns an object which invokes the cleanup function when the object is destroyed. It's used to erase entries from the `std::list<LoadWalletFn>` list when the GUI no longer wants to receive callbacks.

And for that reason a `std::list` is used - iterators aren't invalidated if other elements are added/removed.

fjahr commented:
tested ACK 81ea66c30e2953dee24d5b127c28daa0d9452a28

I was able to reproduce the error using the script from https://github.com/bitcoin/bitcoin/issues/16937 on master but not with this fix.

code review ACK 6d6a7a8403ae923f189812edebdd95761de0e7f2

I was not able to reproduce the duplicate wallet showing up after ~10 tries on master.

Empact commented on src/wallet/wallet.h:
```diff
@@ -48,6 +50,7 @@ bool HasWallets();
 std::vector<std::shared_ptr<CWallet>> GetWallets();
 std::shared_ptr<CWallet> GetWallet(const std::string& name);
 std::shared_ptr<CWallet> LoadWallet(interfaces::Chain& chain, const WalletLocation& location, std::string& error, std::vector<std::string>& warnings);
+std::unique_ptr<interfaces::Handler> HandleLoadWallet(LoadWalletFn load_wallet);
```
nit: this `HandleLoadWallet` arg could be an r-value reference, on account of the `std::move` calling pattern.

kallewoof commented on src/qt/walletframe.h:
```diff
@@ -36,7 +36,7 @@ class WalletFrame : public QFrame
     void setClientModel(ClientModel *clientModel);
 
-    void addWallet(WalletModel *walletModel);
+    bool addWallet(WalletModel *walletModel);
```
I would prefer a `bool ready()` method that simply checks the conditions rather than arbitrarily booling non-bool methods like this. I think it's unnecessarily confusing.

promag replied:
6d6a7a8403ae923f189812edebdd95761de0e7f2

@kallewoof thanks for reviewing.

> arbitrarily booling non-bool methods like this.

What you mean "like this"? I could refactor by adding `bool WalletFrame::hasWallet(WalletModel*) const` but IMO this is also fine, for instance, `std::set::insert` also returns whether the item was added.

What it could have is the `NODISCARD` attribute, but ATM it's not used in src/qt.

Empact replied:
IMO it's an anti-pattern to silently fail when an operation is attempted, returning a success bool instead makes sense to me in that case, or alternatively raise on failure and use `ready` as you suggest.

kallewoof replied:
Yeah, I know this is done elsewhere (even in `std`), but I don't think it's a good idea, personally. It often results in having to look up what exactly is being returned, made even worse with things like `bool CreateThing(...)`. (Looking at the signature unless `Thing` is a bool type, the return value is most likely the result of the create operation, but seeing it in the wild as `auto foo = CreateThing(...)` you won't know if foo is a bool or a `Thing` without checking.)

Agree that failing silently is a bad idea. Raising sounds good to me, unless you really like the bool approach, in which case keep it as is.

kallewoof reviewed (commented):
Code review ACK 6d6a7a8403ae923f189812edebdd95761de0e7f2

laanwj commented:
Code review ACK 6d6a7a8403ae923f189812edebdd95761de0e7f2.

@ryanofsky Thanks for the explanation that was very helpful
